lib/nmatrix/math.rb

```diff
@@ -533,6 +533,77 @@ def solve(b, opts = {})
 
   #
   # call-seq:
+  #     least_squares -> NMatrix
+  #
+  # Provides the linear least squares approximation of an under-determined system  
+  # using QR factorization provided that the matrix is not rank-deficient.
+  #
+  # Only works for dense matrices.
+  #
+  # * *Arguments* :
+  #   - +b+ -> The solution column vector NMatrix of A * X = b.
+  #
+  # * *Returns* :
+  #   - NMatrix that is a column vector with the LLS solution
+  #
+  # * *Raises* :
+  #   - +ArgumentError+ -> least squares approximation only works for non-complex types
+  #   - +ShapeError+ -> system must be under-determined ( rows > columns )
+  #
+  # Examples :-
+  #
+  #   a = NMatrix.new([3,2], [2.0, 0, -1, 1, 0, 2]) 
+  #   
+  #   b = NMatrix.new([3,1], [1.0, 0, -1])
+  #
+  #   a.least_squares(b)
+  #     =>[
+  #         [ 0.33333333333333326 ]
+  #         [ -0.3333333333333334 ]
+  #       ]
+  #
+  def least_squares(b)  
+    raise(ArgumentError, "least squares approximation only works for non-complex types") if 
+      self.complex_dtype?
+    
+    rows, columns = self.shape
+
+    raise(ShapeError, "system must be under-determined ( rows > columns )") unless 
+      rows > columns
+   
+    #Perform economical QR factorization
+    r = self.clone
+    tau = r.geqrf!
+    q_transpose_b = r.ormqr(tau, :left, :transpose, b)
+    
+    #Obtain R from geqrf! intermediate
+    r[0...columns, 0...columns].upper_triangle!
+    r[columns...rows, 0...columns] = 0
+    
+    diagonal = r.diagonal
+
+    raise(ArgumentError, "rank deficient matrix") if diagonal.any? { |x| x == 0 }
+    
+    tolerance = 1e-12
```

Comment:
Arbitrary tolerance here, I'm having trouble figuring out the best way to check if a number is close to zero.